src/Particle/SampleStack.h

```diff
@@ -0,0 +1,91 @@
+//////////////////////////////////////////////////////////////////////////////////////
+// This file is distributed under the University of Illinois/NCSA Open Source License.
+// See LICENSE file in top directory for details.
+//
+// Copyright (c) 2020  QMCPACK developers.
+//
+// File developed by: Mark Dewing, mdewing@anl.gov, Argonne National Laboratory
+//
+// File created by: Mark Dewing, mdewing@anl.gov, Argonne National Laboratory
+//////////////////////////////////////////////////////////////////////////////////////
+
+/** @file SampleStack.h
+ * @brief Stores particle configurations for later use in DMC and wavefunction optimization
+ *
+ * Stores less temporary data than the buffer object.
+ */
+
+#ifndef QMCPLUSPLUS_SAMPLE_STACK_H
+#define QMCPLUSPLUS_SAMPLE_STACK_H
+
+#include <vector>
+#include "Particle/ParticleSet.h"
+#include "Particle/Walker.h"
+
+namespace qmcplusplus
+{
+class MCWalkerConfiguration;
+class HDFWalkerOutput;
+class MCSample;
+
+class SampleStack
+{
+public:
+  typedef ParticleSet::Walker_t Walker_t;
+  typedef std::vector<Walker_t*> WalkerList_t;
+  typedef WalkerList_t::iterator walker_iterator;
+  typedef QMCTraits::PropertySetType PropertySetType;
+
+  SampleStack();
+
+  int setTotalNum(int total_num) { TotalNum = total_num; }
+
+  int getMaxSamples() const { return MaxSamples; }
+
+  void setMaxSamples(int max_samples) { MaxSamples = max_samples; }
+
+  bool empty() const { return sample_stack_.empty(); }
+
+  void putSample(unsigned int i, const Walker_t& w);
+
+  void getSample(unsigned int i, Walker_t& w) const;
+
+  //@{save/load/clear function for optimization
+  inline int getNumSamples() const { return CurSampleCount; }
+  ///set the number of max samples
+  void setNumSamples(int n);
+  ///save the position of current walkers to SampleStack
+  //void saveEnsemble();
+  ///save the position of current walkers
+  void saveEnsemble(walker_iterator first, walker_iterator last);
```

Comment:
Again why create a dependence on the walker type for this.  Perhaps an iterator over std::vector<MCSample&>.

Reply:
That requires writing more code to convert from Walker_t to MCSample.   Which is fine, but it makes the PR larger.  I am trying to make this extraction as minimally intrusive as possible, especially on code outside of MCWalkerConfiguration.